Source/WebCore/html/HTMLAttachmentElement.cpp

```diff
@@ -71,14 +75,103 @@ HTMLAttachmentElement::~HTMLAttachmentElement() = default;
+        ASSERT(attachment->m_implementation == Implementation::Legacy);
+        ASSERT(!attachment->renderer()); // Switch to modern style *must* be done before renderer is created!
+        attachment->m_implementation = Implementation::Modern;
+        attachment->ensureUserAgentShadowRoot();
+    }
+    return attachment;
+}
+
+void HTMLAttachmentElement::didAddUserAgentShadowRoot(ShadowRoot& root)
+{
+    if (m_implementation == Implementation::Modern)
+        ensureModernShadowTree(root);
+}
+
+static const AtomString& attachmentContainerIdentifier()
+{
+    static MainThreadNeverDestroyed<const AtomString> identifier("attachment-container"_s);
+    return identifier;
+}
+
+static const AtomString& attachmentPreviewIdentifier()
+{
+    static MainThreadNeverDestroyed<const AtomString> identifier("attachment-preview"_s);
+    return identifier;
+}
+
+static const AtomString& attachmentTitleIdentifier()
+{
+    static MainThreadNeverDestroyed<const AtomString> identifier("attachment-title"_s);
+    return identifier;
+}
+
+static const AtomString& attachmentSubtitleIdentifier()
+{
+    static MainThreadNeverDestroyed<const AtomString> identifier("attachment-subtitle"_s);
+    return identifier;
 }
 
-RenderPtr<RenderElement> HTMLAttachmentElement::createElementRenderer(RenderStyle&& style, const RenderTreePosition&)
+void HTMLAttachmentElement::ensureModernShadowTree(ShadowRoot& root)
 {
+    ASSERT(m_implementation == Implementation::Modern);
+    if (m_elementWithTitle)
+        return;
+
+    static MainThreadNeverDestroyed<const String> shadowStyle(StringImpl::createWithoutCopying(attachmentElementShadowUserAgentStyleSheet, sizeof(attachmentElementShadowUserAgentStyleSheet)));
+    auto style = HTMLStyleElement::create(HTMLNames::styleTag, document(), false);
+    style->setTextContent(String { shadowStyle });
+    root.appendChild(WTFMove(style));
+
+    auto container = HTMLDivElement::create(document());
+    container->setIdAttribute(attachmentContainerIdentifier());
+    root.appendChild(container);
+
+    // FIXME: This is using the same HTMLAttachmentElement type, but with different behavior (thanks to m_implementation), to fetch and show
+    // the appropriate image (thumbnail, icon, etc.). In the longer term, this functionality should be folded into the Implementation::Modern
+    // code, and the old Legacy/ImageOnly code should be removed. See rdar://105252742.
+    m_innerLegacyAttachment = adoptRef(*new HTMLAttachmentElement(HTMLNames::attachmentTag, document()));
+    m_innerLegacyAttachment->m_implementation = Implementation::ImageOnly;
+    m_innerLegacyAttachment->cloneAttributesFromElement(*this);
+    m_innerLegacyAttachment->m_file = m_file;
+    m_innerLegacyAttachment->m_thumbnail = WTFMove(m_thumbnail);
+    m_innerLegacyAttachment->m_icon = WTFMove(m_icon);
+    m_innerLegacyAttachment->m_iconSize = m_iconSize;
+    m_innerLegacyAttachment->setIdAttribute(attachmentPreviewIdentifier());
```

Comment:
Just wondering — does this mean that we'd have two `API::Attachment` objects in the UI process, one corresponding the outer "modern" attachment and another corresponding to the inner legacy attachment?

Reply:
Like the rest of the shadow DOM elements, the inner attachment element is an implementation detail of how to display icons/thumbnails inside the outer (public) attachment, so it shouldn't be visible anywhere; at least the apps using attachments shouldn't need to know about it.
However I see that *all* attachments call `Document::didInsertAttachmentElement`, and through a series of calls&messages, end in `WebPageProxy::didInsertAttachmentWithIdentifier` where a corresponding `API::Attachment` is created.
I admit I'm not sure what this means. If the UIProcess' `API::Attachment` is needed to display it, then that should be fine and work as expected. Otherwise we could theoretically drop the initial `Document::didInsertAttachmentElement` from inner legacy attachments. I shall explore this further, and I may need to talk to someone knowledgeable to resolve this soon!

Reply:
If you can find a way to extract the code that draws the file icon (without using a nested attachment element), that would be really useful. `<input type="file">` could also use such code to display file icons if we rewrite it to use shadow DOM.

Reply:
Sorry, I should have been clearer: The inner attachment does more than just draw images, it actually requests the proper icon/thumbnail for a given file, it also draws the progress bar; and some of this is different between platforms.
All this should eventually be moved to the modern element layout&rendering, but it was more work&time than using this nesting trick for now -- at least for me!